pyccel/ast/numpyext.py

```diff
@@ -1244,155 +1150,89 @@ def __init__(self,*args):
-        obj = asec(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
-
-#=======================================================================================
-
-class Atan(Function):
-    is_zero = False
-    def __new__(cls,arg):
-        obj = atan(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
-
-
-class Acot(Function):
-    is_zero = False
-    def __new__(cls,arg):
-        obj = acot(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
-
-
-class Acsc(Function):
-    is_zero = False
-    def __new__(cls,arg):
-        obj = acsc(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
-
-#=======================================================================================
-
-class Sinh(Function):
-    is_zero = False
-    def __new__(cls,arg):
-        obj = sinh(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
-
-class Cosh(Function):
-    is_zero = False
-    def __new__(cls,arg):
-        obj = cosh(arg)
-        if arg.is_real:
-            assumptions={'real':True}
-            ass_copy = assumptions.copy()
-            obj._assumptions = StdFactKB(assumptions)
-            obj._assumptions._generator = ass_copy
-        return obj
+#==============================================================================
+# Numpy universal functions
+# https://numpy.org/doc/stable/reference/ufuncs.html#available-ufuncs
+#
+# NOTE: since we are subclassing sympy.Function, we need to use a name ending
+#       with "Base", otherwise the Sympy's printer is going to skip this class.
+#==============================================================================
+class NumpyUfuncBase(Function, PyccelAstNode):
+    """Base class for Numpy's universal functions."""
 
+#------------------------------------------------------------------------------
+class NumpyUfuncUnary(NumpyUfuncBase):
+    """Numpy's universal function with one argument.
+    """
+    def __init__(self, x):
+        self._shape     = x.shape
+        self._rank      = x.rank
+        self._dtype     = x.dtype if x.dtype is NativeComplex() else NativeReal()
```

Comment:
At least some of numpy's functions return the same dtype int the output as the argument. E.g. floor called on an integer returns an integer. I don't know whether floor is the exception or the rule though?

Reply:
floor is the exception, most of the functions returns a real or a complex